lib/galaxy_test/base/decorators.py

```diff
@@ -0,0 +1,101 @@
+"""Test decorators for test methods meant to be run on external Galaxy instances.
+
+These decorators should not be needed for integration tests, unit tests, etc... but
+are appropriate for the API and Selenium tests that are meant to be executable on
+external Galaxies.
+
+Running this class of tests should implicitly come with the expectation that new
+jobs, workflows, and datasets will be created. But tests creating new published objects,
+histories, libraries, etc... should be annotated ideally.
+"""
+import os
+import unittest
+from functools import wraps
+from typing import Union
+
+import pytest
+from typing_extensions import Literal
+
+KnownReqiurementT = Union[
+    Literal["admin"],
+    Literal["new_history"],
+    Literal["new_library"],
+    Literal["new_published_objects"],
+    Literal["new_user"],
+]
```

Comment:
If these were fixtures you could tell the fixture to mark the test, e.g:

```python
@pytest.fixture(marks=pytest.mark.new_history)
def new_history(galaxy_interactor):
    with galaxy_interactor.new_history() as new_history:
        yield new_history
```
guess the question is how much we want to use pytest. My personal preference would be to the max, so we can efficiently split tests based on their fixture use.

Reply:
My approach in this PR provides more static checking and makes more sense for tests that require multiple new histories or for things like ``@requires_new_published_objects`` or ``@requires_admin`` where the thing to inject is not clear at all.

Reply:
I'm not sure what you mean by static checking. Aren't the fixtures providing much more static information ? By static I mean that you can do `--collection-only -m requires_new`, and you don't have to partially run the tests to skip them.

`requires_admin` would be set by the `admin_interactor` fixture, `requires_new_published_objects` would be set by the `create_published_{history,dataset,workflow}` fixture etc. It's a different way of thinking about the test suite for sure, and it'd be a bigger change than what you're doing, but I think it has upsides worth considering ?

Reply:
(I'd still merge the PR, it's a really nice improvement!)

Reply:
The -m pytest markers should work with this approach - I'm marking the methods using pytest inside the decorators. 

```
pytest --collect-only -m "requires_admin" lib/galaxy_test/api/test_roles.py 
===================================================== test session starts =====================================================
platform darwin -- Python 3.9.10, pytest-7.1.3, pluggy-1.0.0 -- /Users/jxc755/workspace/galaxy/.venv/bin/python3.9
cachedir: .pytest_cache
metadata: {'Python': '3.9.10', 'Platform': 'macOS-11.5.2-x86_64-i386-64bit', 'Packages': {'pytest': '7.1.3', 'py': '1.11.0', 'pluggy': '1.0.0'}, 'Plugins': {'playwright': '0.3.0', 'shard': '0.1.2', 'json-report': '1.5.0', 'mock': '3.10.0', 'html': '3.1.1', 'base-url': '2.0.0', 'postgresql': '4.1.1', 'cov': '4.0.0', 'asyncio': '0.20.1', 'anyio': '3.6.2', 'celery': '0.0.0', 'metadata': '2.0.2'}, 'Base URL': ''}
rootdir: /Users/jxc755/workspace/galaxy, configfile: pytest.ini
plugins: playwright-0.3.0, shard-0.1.2, json-report-1.5.0, mock-3.10.0, html-3.1.1, base-url-2.0.0, postgresql-4.1.1, cov-4.0.0, asyncio-0.20.1, anyio-3.6.2, celery-0.0.0, metadata-2.0.2
asyncio: mode=auto
collected 5 items                                                                                                             
Running 5 items in this shard: lib/galaxy_test/api/test_roles.py::TestRolesApi::test_list_and_show, lib/galaxy_test/api/test_roles.py::TestRolesApi::test_create_invalid_params, lib/galaxy_test/api/test_roles.py::TestRolesApi::test_create_valid, lib/galaxy_test/api/test_roles.py::TestRolesApi::test_show_error_codes, lib/galaxy_test/api/test_roles.py::TestRolesApi::test_create_only_admin

<Package api>
  <Module test_roles.py>
    <Class TestRolesApi>
      Base class for tests targetting actual Galaxy servers.
      
      Subclass should override galaxy_driver_class if a Galaxy server
      needs to be launched to run the test, this base class assumes a
      server is already running.
      <Function test_list_and_show>
      <Function test_create_invalid_params>
      <Function test_create_valid>
      <Function test_show_error_codes>
      <Function test_create_only_admin>
```

Invalid decorators, requirements, etc... will be caught by mypy without needing to use pytest is what I meant by more static checking. That means catching problems faster and right in the IDE without needing to run code.

We don't use separate admin and non-admin interactors. We don't structure the code such that create_published_history makes any sense currently. Method decorators are the right place to set method metadata - warping our code in hacky ways for hacky pytest fixtures would be unfortunate.